[GPU Process] [FormControls] Add ControlParts for SearchField and SearchFieldCancelButton

shallawa opened the pull request:
#### d9e61bbb56beb116a6f42f89a97f6ed0adbbb8c0
<pre>
[GPU Process] [FormControls] Add ControlParts for SearchField and SearchFieldCancelButton
<a href="https://bugs.webkit.org/show_bug.cgi?id=249890">https://bugs.webkit.org/show_bug.cgi?id=249890</a>
rdar://103703733

Reviewed by Aditya Keerthi.

These ControlParts will handle drawing the SearchField form control and its cancel
button (i.e. `&lt;input type=&quot;search&quot;&gt;`). For macOS, AppKit will be used to draw the
platform controls.

* Source/WebCore/Headers.cmake:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/spi/mac/NSSearchFieldCellSPI.h: Added.
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/ControlStates.h:
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/SearchFieldCancelButtonPart.h: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
* Source/WebCore/platform/graphics/controls/SearchFieldPart.h: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::ControlFactoryMac::searchFieldCell const):
(WebCore::ControlFactoryMac::createPlatformSearchField):
(WebCore::ControlFactoryMac::createPlatformSearchFieldCancelButton):
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
(WebCore::ControlMac::updateEnabledState):
(WebCore::ControlMac::updateFocusedState):
(WebCore::ControlMac::updatePressedState):
* Source/WebCore/platform/graphics/mac/controls/SearchControlMac.h: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
* Source/WebCore/platform/graphics/mac/controls/SearchControlMac.mm: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
(WebCore::SearchControlMac::SearchControlMac):
(WebCore::SearchControlMac::updateCellStates):
* Source/WebCore/platform/graphics/mac/controls/SearchFieldCancelButtonMac.h: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
* Source/WebCore/platform/graphics/mac/controls/SearchFieldCancelButtonMac.mm: Added.
(WebCore::SearchFieldCancelButtonMac::SearchFieldCancelButtonMac):
(WebCore::SearchFieldCancelButtonMac::cellSize const):
(WebCore::SearchFieldCancelButtonMac::rectForBounds const):
(WebCore::SearchFieldCancelButtonMac::updateCellStates):
(WebCore::SearchFieldCancelButtonMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SearchFieldMac.h: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
* Source/WebCore/platform/graphics/mac/controls/SearchFieldMac.mm: Copied from Source/WebCore/platform/graphics/mac/controls/ControlMac.h.
(WebCore::SearchFieldMac::SearchFieldMac):
(WebCore::SearchFieldMac::draw):
* Source/WebCore/platform/mac/ThemeMac.mm:
(WebCore::ThemeMac::ensuredView):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::createControlPart const):
(WebCore::RenderTheme::extractControlStyleStatesForRenderer const):
(WebCore::RenderTheme::extractControlStyleForRenderer const):
(WebCore::RenderTheme::extractControlStatesForRenderer const):
(WebCore::RenderTheme::isWindowActive const):
(WebCore::RenderTheme::isDefault const):
(WebCore::RenderTheme::isActive const): Deleted.
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::canCreateControlPartForRenderer const):
(WebCore::RenderThemeMac::updatePressedState):
(WebCore::RenderThemeMac::paintSearchField): Deleted.
(WebCore::RenderThemeMac::paintSearchFieldCancelButton): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;ControlPart&gt;::decode):

Canonical link: <a href="https://commits.webkit.org/258462@main">https://commits.webkit.org/258462@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fd8ad080de09d2b50f5a90f430cf4ecec3c0cbaa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102046 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11189 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35113 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111370 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/171560 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106027 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12161 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2101 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109114 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107827 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/92577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/94436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91191 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/24059 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4754 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/25487 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4859 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/1933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/10921 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44977 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5807 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6608 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/PAL/pal/spi/mac/NSSearchFieldCellSPI.h:
```diff
@@ -0,0 +1,32 @@
+
+#import <AppKit/NSSearchFieldCell.h>
+
+@interface NSSearchFieldCell ()
```
The declaration + FIXME in `RenderThemeMac` should be removed.

shallawa replied:
The declaration and  FIXME comment were removed.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
```diff
@@ -52,12 +52,53 @@
     return [[NSWorkspace sharedWorkspace] accessibilityDisplayShouldIncreaseContrast];
 }
 
+void ControlMac::updateEnabledState(NSCell* cell, const ControlStyle& style)
```
Nit: `NSCell *`.

shallawa replied:
Done.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SearchControlMac.mm:
```diff
@@ -0,0 +1,52 @@
+    ASSERT(m_searchFieldCell);
+}
+
+void SearchControlMac::updateCellStates(const FloatRect&, const ControlStyle& style)
```
I think this should call `ControlMac:: updateCellStates` too.

shallawa replied:
You are right. Fixed.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
```diff
@@ -105,6 +108,19 @@
     return m_levelIndicatorCell.get();
 }
 
+NSSearchFieldCell* ControlFactoryMac::searchFieldCell() const
```
Nit: `NSSearchFieldCell *`

shallawa committed:
[GPU Process] [FormControls] Add ControlParts for SearchField and SearchFieldCancelButton
https://bugs.webkit.org/show_bug.cgi?id=249890
rdar://103703733

Reviewed by Aditya Keerthi.

These ControlParts will handle drawing the SearchField form control and its cancel
button (i.e. `<input type="search">`). For macOS, AppKit will be used to draw the
platform controls.

* Source/WebCore/Headers.cmake:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/spi/mac/NSSearchFieldCellSPI.h: Added.
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/ControlStates.h:
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/SearchFieldCancelButtonPart.h: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
* Source/WebCore/platform/graphics/controls/SearchFieldPart.h: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::ControlFactoryMac::searchFieldCell const):
(WebCore::ControlFactoryMac::createPlatformSearchField):
(WebCore::ControlFactoryMac::createPlatformSearchFieldCancelButton):
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
(WebCore::ControlMac::updateEnabledState):
(WebCore::ControlMac::updateFocusedState):
(WebCore::ControlMac::updatePressedState):
* Source/WebCore/platform/graphics/mac/controls/SearchControlMac.h: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
* Source/WebCore/platform/graphics/mac/controls/SearchControlMac.mm: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
(WebCore::SearchControlMac::SearchControlMac):
(WebCore::SearchControlMac::updateCellStates):
* Source/WebCore/platform/graphics/mac/controls/SearchFieldCancelButtonMac.h: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
* Source/WebCore/platform/graphics/mac/controls/SearchFieldCancelButtonMac.mm: Added.
(WebCore::SearchFieldCancelButtonMac::SearchFieldCancelButtonMac):
(WebCore::SearchFieldCancelButtonMac::cellSize const):
(WebCore::SearchFieldCancelButtonMac::rectForBounds const):
(WebCore::SearchFieldCancelButtonMac::updateCellStates):
(WebCore::SearchFieldCancelButtonMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SearchFieldMac.h: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
* Source/WebCore/platform/graphics/mac/controls/SearchFieldMac.mm: Copied from Source/WebCore/platform/graphics/mac/controls/ControlMac.h.
(WebCore::SearchFieldMac::SearchFieldMac):
(WebCore::SearchFieldMac::draw):
* Source/WebCore/platform/mac/ThemeMac.mm:
(WebCore::ThemeMac::ensuredView):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::createControlPart const):
(WebCore::RenderTheme::extractControlStyleStatesForRenderer const):
(WebCore::RenderTheme::extractControlStyleForRenderer const):
(WebCore::RenderTheme::extractControlStatesForRenderer const):
(WebCore::RenderTheme::isWindowActive const):
(WebCore::RenderTheme::isDefault const):
(WebCore::RenderTheme::isActive const): Deleted.
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::canCreateControlPartForRenderer const):
(WebCore::RenderThemeMac::updatePressedState):
(WebCore::RenderThemeMac::paintSearchField): Deleted.
(WebCore::RenderThemeMac::paintSearchFieldCancelButton): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<ControlPart>::decode):

Canonical link: https://commits.webkit.org/258462@main
